clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -116,45 +115,60 @@
     }
 
     public static class Builder extends AbstractRequest.Builder<ProduceRequest> {
-        private final byte magic;
         private final short acks;
         private final int timeout;
         private final Map<TopicPartition, MemoryRecords> partitionRecords;
         private final String transactionalId;
 
-        public Builder(byte magic,
-                       short acks,
-                       int timeout,
-                       Map<TopicPartition, MemoryRecords> partitionRecords,
-                       String transactionalId) {
-            super(ApiKeys.PRODUCE, (short) (magic == RecordBatch.MAGIC_VALUE_V2 ? ApiKeys.PRODUCE.latestVersion() : 2));
-            this.magic = magic;
+        public static Builder forCurrentMagic(short acks,
+                                              int timeout,
+                                              Map<TopicPartition, MemoryRecords> partitionRecords) {
+            return forMagic(RecordBatch.CURRENT_MAGIC_VALUE, acks, timeout, partitionRecords, null);
+        }
+
+        public static Builder forMagic(byte magic,
+                                       short acks,
+                                       int timeout,
+                                       Map<TopicPartition, MemoryRecords> partitionRecords,
+                                       String transactionalId) {
+            // The Produce API is a bit fussy in its version requirements. If the magic version is less than 2,
+            // then we must use version 2; otherwise we want to use the latest version possible as long as it
+            // is greater than 2.
```

Comment:
The first sentence is fine. The rest seems to just repeat the code?

Reply:
Yes, you are right. I think it is worth explaining here why we must use the old produce version for older message format versions.